README.md
# react-native-dialog

[![npm version](https://badge.fury.io/js/react-native-dialog.svg)](https://badge.fury.io/js/react-native-dialog)

A flexible pure JavaScript React-Native dialog that follows closely the native UI guidelines.

## Features

- Support for iOS and Android (JavaScript API)
- A flexible declarative API
- Follows closely the UI of native dialogs/alerts
- Can be used both as an alert and as an input prompt
- Can be injected with any component
- Supports light/dark mode

## Demo

<p align="center">
<img src="https://raw.githubusercontent.com/mmazzarolo/react-native-dialog/master/.github/docs-images/react-native-dialog-ios-alert.png" height="500" />
<img src="https://raw.githubusercontent.com/mmazzarolo/react-native-dialog/master/.github/docs-images/react-native-dialog-android-alert.png" height="500" />
</p>

<p align="center">
<img src="https://raw.githubusercontent.com/mmazzarolo/react-native-dialog/master/.github/docs-images/react-native-dialog-ios-input.png" height="500" />
<img src="https://raw.githubusercontent.com/mmazzarolo/react-native-dialog/master/.github/docs-images/react-native-dialog-android-input.png" height="500" />
</p>

## Setup

Install the library using npm or yarn:

```bash
# Using npm:
$ npm install react-native-dialog
# Using yarn:
$ yarn add react-native-dialog
```

## Usage

React-native-dialog exposes a set of components that can be used to build the UI of the dialog:

- **Dialog.Container**: This component is the root component of the dialog and all the other components should be nested inside it.
- **Dialog.Title**: A `Text` component styled as a native dialog title.
- **Dialog.Description**: A `Text` component styled as a native dialog description.
- **Dialog.Button**: A component styled as a native dialog button.
- **Dialog.Input**: A `TextInput` component styled as a native dialog input.
- **Dialog.CodeInput**: A `TextInput` component styled as one time code input.
- **Dialog.Switch**: A native `Switch` component with an optional label.

1. Import react-native-dialog:

```javascript
import Dialog from "react-native-dialog";
```

2. Create a dialog and nest its content inside of it:

```javascript
return (
  <View>
    <Dialog.Container>
      <Dialog.Title>Account delete</Dialog.Title>
      <Dialog.Description>
        Do you want to delete this account? You cannot undo this action.
      </Dialog.Description>
      <Dialog.Button label="Cancel" />
      <Dialog.Button label="Delete" />
    </Dialog.Container>
  </View>
);
```

3. Then simply show it by setting the `visible` prop to true:

```javascript
return (
  <View>
    <Dialog.Container visible={true}>
      <Dialog.Title>Account delete</Dialog.Title>
      <Dialog.Description>
        Do you want to delete this account? You cannot undo this action.
      </Dialog.Description>
      <Dialog.Button label="Cancel" />
      <Dialog.Button label="Delete" />
    </Dialog.Container>
  </View>
);
```

The `visible` prop is the only prop you'll really need to make the dialog work: you should control this prop value by saving it in your state and setting it to `true` or `false` when needed.

## A complete example

The following example consists in a component (`DialogTester`) with a button and a dialog.
The dialog is controlled by the `dialogVisible` state variable and it is initially hidden since its value is `false`.
Pressing the button sets `dialogVisible` to true, making the dialog visible.
Inside the dialog there are two buttons that, when pressed, set `dialogVisible` to false, hiding the dialog.

```javascript
import React, { useState } from "react";
import { Button, StyleSheet, View } from "react-native";
import Dialog from "react-native-dialog";

export default function App() {
  const [visible, setVisible] = useState(false);

  const showDialog = () => {
    setVisible(true);
  };

  const handleCancel = () => {
    setVisible(false);
  };

  const handleDelete = () => {
    // The user has pressed the "Delete" button, so here you can do your own logic.
    // ...Your logic
    setVisible(false);
  };

  return (
    <View style={styles.container}>
      <Button title="Show dialog" onPress={showDialog} />
      <Dialog.Container visible={visible}>
        <Dialog.Title>Account delete</Dialog.Title>
        <Dialog.Description>
          Do you want to delete this account? You cannot undo this action.
        </Dialog.Description>
        <Dialog.Button label="Cancel" onPress={handleCancel} />
        <Dialog.Button label="Delete" onPress={handleDelete} />
      </Dialog.Container>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",
  },
});
```

## Available props

### Dialog.Button props

| Name     | Type   | Default                                | Description                             |
| -------- | ------ | -------------------------------------- | --------------------------------------- |
| label    | string | **REQUIRED**                           | The label text                          |
| color    | string | `#007ff9` on iOS, `#169689` on Android | The label color                         |
| bold     | bool   | false                                  | Show the label with a bold font weight? |
| disabled | bool   | false                                  | Disable the button?                     |
| onPress  | func   | **REQUIRED**                           | Called when the button is pressed       |

### Dialog.Description props

| Name     | Type   | Default      | Description          |
| -------- | ------ | ------------ | -------------------- |
| children | string | **REQUIRED** | The description text |

### Dialog.Container props

| Name                   | Type   | Default       | Description                                                                                         |
| ---------------------- | ------ | ------------- | --------------------------------------------------------------------------------------------------- |
| blurComponentIOS       | node   | A low-opacity | The blur component used in iOS                                                                      |
| visible                | bool   | **REQUIRED**  | Show the dialog?                                                                                    |
| children               | node   | **REQUIRED**  | The dialog content                                                                                  |
| contentStyle           | any    | undefined     | Extra style applied to the dialog content                                                           |
| headerStyle            | any    | undefined     | Extra style applied to the dialog header                                                            |
| footerStyle            | any    | undefined     | Extra style applied to the dialog footer                                                            |
| buttonSeparatorStyle   | any    | undefined     | Extra style applied to the dialog button separator                                                  |
| onBackdropPress        | func   | undefined     | Callback invoked when the backdrop is pressed                                                       |
| onRequestClose         | func   | undefined     | Callback invoked when the hardware back button on Android or the menu button on Apple TV is pressed |
| keyboardVerticalOffset | number | undefined     | keyboardVerticalOffset for iOS                                                                      |
| verticalButtons        | bool   | false         | Renders button vertically                                                                           |
| useNativeDriver        | bool   | false         | Defines if animations should use native driver                                                      |

### Dialog.Input props

| Name               | Type   | Default   | Description                                                             |
| ------------------ | ------ | --------- | ----------------------------------------------------------------------- |
| label              | string | undefined | The input floating label                                                |
| wrapperStyle       | any    | undefined | The style applied to the input wrapper View                             |
| textInputRef       | ref    | undefined | Ref to the input                                                        |
| unstableLabelStyle | any    | undefined | Likely to be removed in a future version. See issue #141 for discussion |

`Dialog.Input` also accepts all the React-Native's `TextInput` component props.

### Dialog.CodeInput props

| Name                       | Type   | Default   | Description                                                 |
| -------------------------- | ------ | --------- | ----------------------------------------------------------- |
| wrapperStyle               | any    | undefined | The style applied to the input wrapper View                 |
| digitContainerStyle        | any    | undefined | The style applied to the digit container View               |
| digitContainerFocusedStyle | any    | undefined | The style applied to the digit container View when in focus |
| digitStyle                 | any    | undefined | The style applied to the digit text                         |
| codeLength                 | number | 4         | The total number of digits                                  |
| onCodeChange               | func   | undefined | Called when the input changed                               |

`Dialog.CodeInput` also accepts all the React-Native's `TextInput` component props.

### Dialog.Title props

| Name     | Type   | Default      | Description    |
| -------- | ------ | ------------ | -------------- |
| children | string | **REQUIRED** | The title text |

`Dialog.Title` also accepts all the React-Native's `Text` component props.

### Dialog.Switch props

| Name               | Type   | Default   | Description                                                             |
| ------------------ | ------ | --------- | ----------------------------------------------------------------------- |
| label              | string | undefined | The switch description text                                             |
| unstableLabelStyle | any    | undefined | Likely to be removed in a future version. See issue #141 for discussion |

`Dialog.Switch` also accepts all the React-Native's `Switch` component props.

## Frequently Asked Questions

### How can I use a custom blur component as the dialog background on iOS?

To achieve a look even closer to the native iOS dialog you can provide your own component in the `blurComponentIOS` prop of a `Dialog.Container` and it will be injected in the dialog to be used as a background.
The `blurComponentIOS` can be useful for example if you want to apply a native blur effect to the dialog.
Here is an example using `react-native-blur`:

```javascript
const blurComponentIOS = (
  <BlurView style={StyleSheet.absoluteFill} blurType="xlight" blurAmount={50} />
);
return (
  <View style={styles.container}>
    <Dialog.Container visible={visible} blurComponentIOS={blurComponentIOS}>
      <Dialog.Title>Account delete</Dialog.Title>
      <Dialog.Description>
        Do you want to delete this account? You cannot undo this action.
      </Dialog.Description>
      <Dialog.Button label="Cancel" onPress={handleCancel} />
      <Dialog.Button label="Delete" onPress={handleConfirm} />
    </Dialog.Container>
  </View>
);
```

### How can I add a 'tap outside dialog' callback?

`react-native-dialog` uses [a thin abstraction on top of the React-Native's modal component](./src/Modal.tsx). Any properties you add to `Dialog.Container` are mapped through to the modal.
The modal has an `onBackdropPress` property that can be used to register clicks on the backdrop.

Below is an example of how you can close the dialog by tapping outside.

```javascript
const [visible, setVisible] = useState(true);

const handleCancel = () => {
  setVisible(false);
};

return (
  <Dialog.Container visible={visible} onBackdropPress={handleCancel}>
    <Dialog.Title>Title</Dialog.Title>
    <Dialog.Button label="Cancel" onPress={handleCancel} />
  </Dialog.Container>
);
```

## Acknowledgments

Thanks to the user [@honaf](https://github.com/honaf) who has kindly offered the `react-native-dialog` namespace.
Also thanks to the user [@leecade](https://github.com/leecade) who offered the namespace `react-native-alert` (which has not been used since "Dialog" seems to suit better this component) and to [@tyxou](https://github.com/tyxou) for the entire codebase refactoring to hooks.


CHANGELOG.md
# [9.3.0](https://github.com/mmazzarolo/react-native-dialog/compare/v9.2.2...v9.3.0) (2022-10-31)


### Features

* Add unstableLabelStyle & fix switch statement comparison method ([#150](https://github.com/mmazzarolo/react-native-dialog/issues/150)) ([839915b](https://github.com/mmazzarolo/react-native-dialog/commit/839915bc546707526f3853d454fbe87b24d2cac3))

## [9.2.2](https://github.com/mmazzarolo/react-native-dialog/compare/v9.2.1...v9.2.2) (2022-06-09)


### Bug Fixes

* remove usage of PropTypes ([#139](https://github.com/mmazzarolo/react-native-dialog/issues/139)) ([538ff00](https://github.com/mmazzarolo/react-native-dialog/commit/538ff007942145d1d9aa85a34d65afa25015db16))

## [9.2.1](https://github.com/mmazzarolo/react-native-dialog/compare/v9.2.0...v9.2.1) (2022-02-11)


### Bug Fixes

* ios description platform color ([#133](https://github.com/mmazzarolo/react-native-dialog/issues/133)) ([2952f47](https://github.com/mmazzarolo/react-native-dialog/commit/2952f470130ce119e5d5dcd4844d3bcae3513e3d))

# [9.2.0](https://github.com/mmazzarolo/react-native-dialog/compare/v9.1.2...v9.2.0) (2021-10-23)


### Features

* Add `CodeInput` component ([#128](https://github.com/mmazzarolo/react-native-dialog/issues/128)) ([961f9a0](https://github.com/mmazzarolo/react-native-dialog/commit/961f9a0e9c1803fcc1bfcd8ded1e697054899bb9))

## [9.1.2](https://github.com/mmazzarolo/react-native-dialog/compare/v9.1.1...v9.1.2) (2021-10-01)


### Bug Fixes

* Change style propType of Input from View to Text ([#124](https://github.com/mmazzarolo/react-native-dialog/issues/124)) ([d71d204](https://github.com/mmazzarolo/react-native-dialog/commit/d71d20489c5d4e7038ebbb8dc183fc352d87b409))

## [9.1.1](https://github.com/mmazzarolo/react-native-dialog/compare/v9.1.0...v9.1.1) (2021-09-30)


### Bug Fixes

* Make Button label prop type required ([#123](https://github.com/mmazzarolo/react-native-dialog/issues/123)) ([e096d6b](https://github.com/mmazzarolo/react-native-dialog/commit/e096d6bf6814a114bb635585caa0babe72c3bd1c))

# [9.1.0](https://github.com/mmazzarolo/react-native-dialog/compare/v9.0.0...v9.1.0) (2021-09-08)


### Features

* Expose `useNativeDriver` as `Container` prop ([#122](https://github.com/mmazzarolo/react-native-dialog/issues/122)) ([dd6d979](https://github.com/mmazzarolo/react-native-dialog/commit/dd6d979279b22ad869d95ba530a1e9cd1fc1601b))

# [9.0.0](https://github.com/mmazzarolo/react-native-dialog/compare/v8.2.0...v9.0.0) (2021-09-08)


### Bug Fixes

* Typescript updates ([#121](https://github.com/mmazzarolo/react-native-dialog/issues/121)) ([1ba6f16](https://github.com/mmazzarolo/react-native-dialog/commit/1ba6f16fb2127f8da7e41c79444709b56cb0d4f6))


### BREAKING CHANGES

* Updated TypeScript type definitions. Types _should_ just be more relaxed now — but since we updated the entire codebase we'll release this a major bump to be safe.

# [8.2.0](https://github.com/mmazzarolo/react-native-dialog/compare/v8.1.1...v8.2.0) (2021-07-19)


### Features

* Handle hardware back button on Android ([#114](https://github.com/mmazzarolo/react-native-dialog/issues/114)) ([e9b6cf5](https://github.com/mmazzarolo/react-native-dialog/commit/e9b6cf583cb5070cbb4c542ec9569e29fae3877a))

## [8.1.1](https://github.com/mmazzarolo/react-native-dialog/compare/v8.1.0...v8.1.1) (2021-07-15)


### Bug Fixes

* **title:** children prop allows node instead of string ([#112](https://github.com/mmazzarolo/react-native-dialog/issues/112)) ([8290768](https://github.com/mmazzarolo/react-native-dialog/commit/8290768e73cf22e85404c6ebec9a328a3f0bf022))

# [8.1.0](https://github.com/mmazzarolo/react-native-dialog/compare/v8.0.1...v8.1.0) (2021-07-12)


### Features

* Added support for vertical buttons ([bebc3d0](https://github.com/mmazzarolo/react-native-dialog/commit/bebc3d040bdc0749e5bfbdc3c05ceebfaec7c8d5))

## [8.0.1](https://github.com/mmazzarolo/react-native-dialog/compare/v8.0.0...v8.0.1) (2021-04-18)


### Bug Fixes

* Mark "textInputRef" and "blurComponentIOS" as optional ([#103](https://github.com/mmazzarolo/react-native-dialog/issues/103)) ([c92fe5d](https://github.com/mmazzarolo/react-native-dialog/commit/c92fe5d1492bbe55d06c7af4dec2be653360640f))

# [8.0.0](https://github.com/mmazzarolo/react-native-dialog/compare/v7.0.0...v8.0.0) (2021-04-17)


### Features

* Use TypeScript in the codebase and expose TypeScript type-definitions ([#101](https://github.com/mmazzarolo/react-native-dialog/issues/101)) ([77d41f6](https://github.com/mmazzarolo/react-native-dialog/commit/77d41f6f5fae17650245684c10ab3de3df93e76b))


### BREAKING CHANGES

* Previous type definitions used in your app may not be compatible with this release anymore.

# [7.0.0](https://github.com/mmazzarolo/react-native-dialog/compare/v6.2.0...v7.0.0) (2021-04-16)


### Features

* Add support for dark mode ([#100](https://github.com/mmazzarolo/react-native-dialog/issues/100)) ([57b065e](https://github.com/mmazzarolo/react-native-dialog/commit/57b065e1524e64f28b7a07ebd8062d7b1982cc76))


### BREAKING CHANGES

* We are now using the native colors instead of the hardcoded ones — which means that if your react-native app is using a specific accent/primary color it will be now shown correctly in the dialog.

# [6.2.0](https://github.com/mmazzarolo/react-native-dialog/compare/v6.1.2...v6.2.0) (2021-02-19)


### Features

* add keyboardVerticalOffset support for iOS ([#93](https://github.com/mmazzarolo/react-native-dialog/issues/93)) ([f44d21b](https://github.com/mmazzarolo/react-native-dialog/commit/f44d21bbe72183c129fba72b79440af26c348b1e))

## [6.1.2](https://github.com/mmazzarolo/react-native-dialog/compare/v6.1.1...v6.1.2) (2020-10-24)


### Bug Fixes

* Always cover the whole screen (even with translucent status bar)  ([#85](https://github.com/mmazzarolo/react-native-dialog/issues/85)) ([09b2f35](https://github.com/mmazzarolo/react-native-dialog/commit/09b2f3584890be76fd56d3e2719ea928e8130ebf))


src/Container.js
import PropTypes from "prop-types";
import React from "react";
import {
  KeyboardAvoidingView,
  Platform,
  StyleSheet,
  View,
  PlatformColor,
} from "react-native";
import Modal from "./Modal";
import useTheme from "./useTheme";

const iOS = Platform.OS === "ios";

const DialogContainer = (props) => {
  const {
    blurComponentIOS,
    buttonSeparatorStyle = {},
    children,
    contentStyle = {},
    footerStyle = {},
    headerStyle = {},
    blurStyle = {},
    visible,
    keyboardVerticalOffset = 40,
    ...nodeProps
  } = props;
  const titleChildrens = [];
  const descriptionChildrens = [];
  const buttonChildrens = [];
  const otherChildrens = [];
  const { styles } = useTheme(buildStyles);
  React.Children.forEach(children, (child) => {
    if (!child) {
      return;
    }
    if (
      child.type.name === "DialogTitle" ||
      child.type.displayName === "DialogTitle"
    ) {
      titleChildrens.push(child);
    } else if (
      child.type.name === "DialogDescription" ||
      child.type.displayName === "DialogDescription"
    ) {
      descriptionChildrens.push(child);
    } else if (
      child.type.name === "DialogButton" ||
      child.type.displayName === "DialogButton"
    ) {
      if (Platform.OS === "ios" && buttonChildrens.length > 0) {
        buttonChildrens.push(
          <View style={[styles.buttonSeparator, buttonSeparatorStyle]} />
        );
      }
      buttonChildrens.push(child);
    } else {
      otherChildrens.push(child);
    }
  });
  return (
    <Modal
      renderToHardwareTextureAndroid={true}
      transparent={true}
      visible={visible}
      {...nodeProps}
    >
      <KeyboardAvoidingView
        behavior={iOS ? "padding" : undefined}
        keyboardVerticalOffset={iOS ? keyboardVerticalOffset : undefined}
        style={styles.centeredView}
      >
        <View style={[styles.content, contentStyle]}>
          {Platform.OS === "ios" && blurComponentIOS}
          {Platform.OS === "ios" && !blurComponentIOS && (
            <View style={[styles.blur, blurStyle]} />
          )}
          <View style={[styles.header, headerStyle]}>
            {titleChildrens}
            {descriptionChildrens}
          </View>
          {otherChildrens}
          {Boolean(buttonChildrens.length) && (
            <View style={[styles.footer, footerStyle]}>
              {buttonChildrens.map((x, i) =>
                React.cloneElement(x, {
                  key: `dialog-button-${i}`,
                })
              )}
            </View>
          )}
        </View>
      </KeyboardAvoidingView>
    </Modal>
  );
};

DialogContainer.propTypes = {
  blurComponentIOS: PropTypes.node,
  buttonSeparatorStyle: PropTypes.object,
  children: PropTypes.node.isRequired,
  contentStyle: PropTypes.object,
  footerStyle: PropTypes.object,
  headerStyle: PropTypes.object,
  blurStyle: PropTypes.object,
  visible: PropTypes.bool,
  onBackdropPress: PropTypes.func,
};

DialogContainer.defaultProps = {
  visible: false,
};

const buildStyles = (isDark) =>
  StyleSheet.create({
    centeredView: {
      marginTop: 22,
    },
    blur: {
      position: "absolute",
      backgroundColor: PlatformColor("systemGray6"), // "rgb(255,255,255)",
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
    },
    content: Platform.select({
      ios: {
        width: 270,
        //backgroundColor: PlatformColor("systemGray6"),
        flexDirection: "column",
        borderRadius: 13,
        overflow: "hidden",
      },
      android: {
        backgroundColor: PlatformColor("?attr/colorBackgroundFloating"),
        flexDirection: "column",
        borderRadius: 3,
        padding: 16,
        margin: 16,
        overflow: "hidden",
        elevation: 4,
        minWidth: 300,
      },
      web: {
        flexDirection: "column",
        borderRadius: 3,
        padding: 16,
        margin: 16,
        backgroundColor: "white",
        overflow: "hidden",
        elevation: 4,
        minWidth: 300,
      },
    }),
    header: Platform.select({
      ios: {
        margin: 18,
      },
      android: {
        margin: 12,
      },
      web: {
        margin: 12,
      },
    }),
    footer: Platform.select({
      ios: {
        flexDirection: "row",
        justifyContent: "space-between",
        borderTopColor: PlatformColor("separator"), //"#A9ADAE",
        borderTopWidth: StyleSheet.hairlineWidth,
        height: 46,
      },
      android: {
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "flex-end",
        marginTop: 4,
      },
      web: {
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "flex-end",
        marginTop: 4,
      },
    }),
    buttonSeparator: {
      height: "100%",
      backgroundColor: PlatformColor("separator"), //"#A9ADAE",
      width: StyleSheet.hairlineWidth,
    },
  });

export default DialogContainer;
